pyccl/parameters.py

```diff
@@ -0,0 +1,120 @@
+from . import ccllib as lib
+from .errors import warnings, CCLDeprecationWarning
+
+
+class CCLParameters:
+    """Base for classes holding global CCL parameters and their values.
+
+    Subclasses contain a reference to the C-struct with the collection
+    of parameters and their values (via SWIG). All subclasses act as proxies
+    to the CCL parameters at the C-level.
+
+    Subclasses automatically store a backup of the initial parameter state
+    to enable ad hoc reloading.
+    """
+
+    def __init_subclass__(cls, instance=None, freeze=False):
+        """Routine for subclass initialization.
+
+        Parameters:
+            instance (``pyccl.ccllib.cvar``):
+                Reference to ``cvar`` where the parameters are implemented.
+            freeze (``bool``):
+                Disable parameter mutation.
+        """
+        super().__init_subclass__()
+        cls._instance = instance
+        cls._frozen = freeze
+
+        def _new_setattr(self, key, value):
+            # Make instances of the SWIG-level class immutable
+            # so that everything is handled through this interface.
+            # SWIG only assigns `this` via the low level `_ccllib`;
+            # we therefore disable all other direct assignments.
+            if key == "this":
+                return object.__setattr__(self, key, value)
+            name = self.__class__.__name__
+            # TODO: Deprecation cycle for fully immutable Cosmology objects.
+            # raise AttributeError(f"Direct assignment in {name} not supported.")  # noqa
+            warnings.warn(
+                f"Direct assignment of {name} is deprecated. "
+                f"Set via `pyccl.{name}.{key}` before instantiation.",
+                CCLDeprecationWarning)
+            object.__setattr__(self, key, value)
+
+        cls._instance.__class__.__setattr__ = _new_setattr
+
+    def __init__(self):
+        # Keep a copy of the default parameters.
+        object.__setattr__(self, "_bak", CCLParameters.get_params_dict(self))
```

Comment:
Just so I understand: here you're using `object.__setattr__` instead of `self._bak` because you changed `self`'s setattr below, right?

Reply:
Yes. Before `__init__` is called, `self`'s dictionary doesn't contain the entry `_bak`. So if you try to do `self._bak` directly, it will `raise KeyError(f"Parameter _bak does not exist.")`.